File: src/leetcode/java/medium/CoinChange_322.java
package leetcode.java.medium;

import java.util.Arrays;

public class CoinChange_322 {
  public int coinChange(int[] coins, int amount) {
    if (amount == 0) return 0;
    int[] dp = new int[amount + 1];
    Arrays.fill(dp, -1);
    dp[0] = 0;
    for (int coin : coins) {
      for (int i = coin; i <= amount; i++) {
        if (dp[i - coin] != -1) {
          if (dp[i] == -1) dp[i] = dp[i - coin] + 1;
          else dp[i] = Math.min(dp[i], dp[i - coin] + 1);
        }
      }
    }
    return dp[amount];
  }
}

File: src/leetcode/python/medium/CoinChange_322.py
class CoinChange_322(object):
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        if amount == 0:
            return 0
        dp = [-1] * (amount+1)
        dp[0] = 0
        for c in coins:
            for i in range(c, amount+1):
                if dp[i-c] != -1:
                    if dp[i] == -1:
                        dp[i] = dp[i-c] + 1
                    else:
                        dp[i] = min(dp[i-c]+1, dp[i])
        return dp[amount]
